### ISSUE TITLE
Implement string interpolation

### ISSUE THREAD

**Issue by jfecher** [jfecher/ante#91]
String interpolation is the last feature from the old C++ compiler that is still unimplemented in the rust compiler.

Once implemented this feature should desugar strings in the form of `"... ${expr} ..."` into `"... " ++ cast expr : string ++ " ..."`. This looks like the following in the Ast:
```rs
FunctionCall(Variable("++"), vec![
    StringLiteral("... "),
    FunctionCall(Variable("++"), vec![
        TypeAnnotation(
            FunctionCall(Variable("cast"), vec![expr]),
            Type::String,
        ),
        StringLiteral(" ..."),
    ])
])
```

The first and last parts of the string can be elided if they're empty. We may also want to desugar strings in the form `"... $identifier ..."` to work equivalently to `"... ${identifier} ...`.

An implementation of this feature should likely perform this desugaring when constructing the string node while parsing - checking if the string contains `$` and so on. We'll also need to allow escaping `$` via `\`. As a design note, it may be that the `" $identifier "` syntax collides too much with users expecting to use `$` characters in the string. If this is the case only allowing the `${expr}` syntax may be preferable.

It may be in the future we want a more advanced desugaring so that we can e.g. desugar to some `LazyInterpolate` type to avoid constructing strings that will be immediately logged. This is an optimization that can wait.

**Comment by ehllie** [comment]
Should all strings be implicitly formatted like this, or should we do it like python does it for example, and only interpolate strings prefixed with `f` or using some other identifier?
```python
foo = "Hello"
"{foo}, world!" # "{foo}, world!"
f"{foo}, world!" # "Hello, world!"
```

**Comment by ehllie** [comment]
Here's a couple examples of how other languages have done this:

### Python
Like i mentioned above, python uses fstring format, where you can interpolate any expression inside a string prefixed with `f` by wrapping it in `{expr}`
```python
foo = lambda x: x + ", World"
bar = "Hello"
print(f"{foo(bar)}!")  # Hello, World!
```

### JavaScript
JavaScript uses backticks, and then indicates expressions to interpolate with a `${expr}` syntax.
```js
foo = (x) => x + ', World'
bar = 'Hello'
console.log(`${foo(bar)}!`) // Hello, World!
```

### Rust
Rust uses macros to achieve string interpolation, and the expressions to interpolate are passed in as arguments to the macro
```rust
fn main() {
    fn foo(x: &str) -> String {
        let mut ret = String::from(x);
        ret.push_str(", World");
        ret
    }
    let bar = "Hello";
    println!("{}!", foo(bar)); // Hello, World!
}
```

### Bash
Bash interpolates any `""`, but leaves `''` unchanged, making it similar to JavaScript in that aspect.
```bash
function foo {
  echo "$1, World"
}
bar="Hello"
echo "$(foo $bar)!" # Hello, World!
```

***

In my opinion, there should be a way of expressing a string that is not meant to be interpolated, without needing to escape any characters that are used for interpolation. The bikeshedding issue however, is what the syntax for interpolated vs non interpolated strings should be. I find python's `f"{expr}"` syntax most elegant because that's what I'm most used to, and because it requires no additional tokens beyond curly braces inside the string to indicate where the expressions start and begin. JavaScript's approach of `` `${expr}` `` is is a little to verbose in my opinion, because you not only indicate an interpolated string with, but also then have to indicate an expression with both curly brackets and a `$`.

Beyond that, the function syntax should be achievable from the stdlib, with something like so
```scala
trait Format t with
  format: string - t -> Maybe string
```

**Comment by jfecher** [comment]
My thinking is that `"... ${expr} ..."` is a rare enough sequence of tokens such that we won't need two different string forms to exclude string interpolation. If we are worried about this however, I am still open to swift's syntax which uses `"... \(expr) ..."`. The advantage here being that `\(` is an invalid escape sequence so this syntax never actually conflicts with any valid string's syntax.
I think this approach is better than introducing a second string form for non-escaped strings since the later sounds like unnecessary complexity with rather minimal difference. That being said, a second string form for completely raw strings with no escapes at all (or allowing for different string starters/terminators other than quotes) is still on the table in the future.

**Comment by jfecher** [comment]
As for the proposed `Format` trait it seems to provide two things over the `... ++ cast expr : string ++ ...` approach:
1. A somewhat higher efficiency: no need for an intermediate string for `expr`, may be able to preallocate all the capacity needed in a resulting string beforehand.
2. A fallible format by returning `None`.

I think the convenience of defining a to string conversion via `impl Cast T string` and also getting string interpolation for free may be better than a slight efficiency gain for now at least while ante is immature. In the future there are also other avenues for efficiency, e.g. in large strings with many format specifiers the `Format` trait will still lead to many intermediate strings and concatenations. We could explore options like a `Emit string` effect being handled to collect all of the strings being interpolated and allocate a single result string with the total length of all of them. There are alternatives without effects as well.
As for fallible conversions, it isn't clear to me why these may be desirable. My thinking is that any central format should be infallible - save some exceptional cases like running out of memory. I'll have to look more into the error cases of rust's `std::fmt::Result` though.

**Comment by ehllie** [comment]
My reasoning for bringing up a `Format` trait that can fail, was to enable things like using an `Iterator` that gets passed at runtime to perform string interpolation. There would be no guarantee that the `Iterator` has the correct number of items inside it to perform the interpolation, therefore it could sometimes fail due that.

**Comment by jfecher** [comment]
Implemented in #131
